kernel/include/kthread.h

```diff
@@ -204,6 +204,13 @@ static ALWAYS_INLINE bool should_preempt(struct k_thread *thread,
 		return true;
 	}
 
+	/* Otherwise we have to be running a preemptible thread or
+	 * switching to a metairq
+	 */
+	if (thread_is_preemptible(_current) || thread_is_metairq(thread)) {
+		return true;
+	}
```

Comment:
Worth pointing out that if we fix swap() to remove PendSV, we get this first patch's effect for free as the code you're reordering it around would disappear.

Reply:
Oh, and I'm going to bet that the real effect is that the reordering isn't about _current at all, it's that the two cases you're moving are much more common than the SWAP_NONATOMIC edge case, so are effecting an early exit that doesn't need to check z_is_thread_timeout_active().